src/Language/DifferentialDatalog/Debugger/DebugEventParser.hs

```diff
@@ -0,0 +1,160 @@
+
+parens :: ParsecT String u Identity a -> ParsecT String u Identity a
+parens       = T.parens lexer
+
+brackets :: ParsecT String u Identity a -> ParsecT String u Identity a
+brackets     = T.brackets lexer
+
+decimal :: ParsecT String u Identity Integer
+decimal      = T.decimal lexer
+
+double :: ParsecT String u Identity Double
+double       = T.float lexer
+
+dot :: ParsecT String u Identity String
+dot          = T.dot lexer
+
+stringLit :: ParsecT String u Identity String
+stringLit    = T.stringLiteral lexer
+
+
+operatorParser :: ParsecT String u Identity Operator
+operatorParser = do
+                    op <- identifier
+                    case op of
+                        "Map" -> return OpMap
+                        "Aggregate" -> return OpAggregate
+                        "Condition" -> return OpCondition
+                        "Join" -> return OpJoin
+                        "Antijoin" -> return OpAntijoin
+                        "Inspect" -> return OpInspect
+                        _ -> return  OpUndefined
+
+
+intParser :: ParsecT String u Identity Record
+intParser = IntRecord <$> decimal
+
+boolParser :: ParsecT String u Identity Record
+boolParser = BoolRecord <$> ((symbol "true" >> return True)
+                              <|> (symbol "false" >> return False))
+
+doubleParser :: ParsecT String u Identity Record
+doubleParser = DoubleRecord <$> double
+
+stringParser :: ParsecT String u Identity Record
+stringParser = StringRecord <$> stringLit
+
+namedStructParser :: ParsecT String u Identity Record
+namedStructParser = NamedStructRecord <$> identifier <*> (braces (commaSep ((,) <$> (dot *> identifier) <*> (symbol "=" *> recordParser))))
+
+tupleParser :: ParsecT String u Identity Record
+tupleParser = parens (TupleRecord <$> (commaSep recordParser))
+
+arrayParser :: ParsecT String u Identity Record
+arrayParser =  brackets (ArrayRecord Vector <$> (commaSep recordParser))
+
+recordParser :: ParsecT String u Identity Record
+recordParser = try (doubleParser) <|>boolParser <|> intParser <|> stringParser
+               <|> namedStructParser <|> tupleParser <|> arrayParser
+
+debugEventParser :: ParsecT String u Identity Event
+debugEventParser = DebugEvent <$> parens ((,,) <$> (decimal <* comma) <*> (decimal <* comma) <*> decimal)
+                              <*> (comma *> decimal) <*> (comma *> decimal) <*> (comma *> operatorParser)
+                              <*> (comma *> recordParser) <*> (comma *> recordParser)
+
+debugJoinEventParser :: ParsecT String u Identity Event
+debugJoinEventParser = DebugJoinEvent <$> parens ((,,) <$> (decimal <* comma) <*> (decimal <* comma) <*> decimal)
+                              <*> (comma *> decimal) <*> (comma *> decimal) <*> (comma *> operatorParser)
+                              <*> (comma *> recordParser) <*> (comma *> recordParser) <*> (comma *> recordParser)
+
+eventParser :: ParsecT String u Identity Event
+eventParser = debugEventParser <|> debugJoinEventParser
```

Comment:
Parsec does not do lookahead automatically, so this rule will always run `debugEvenParser`, which will successfully consume the first few fields, and then fail if this is a join event. Try this approach instead:

```
debugEventParser = do
    opid <- parens ((,,) <$> (decimal <* comma) <*> (decimal <* comma) <*> decimal)
    w <- comma *> decimal
    ts <- comma *> decimal <* comma
    op <- operatorParser
    case op of
             OpJoin -> ... -- parse three records
                              return DebugJoinEvent{ ....}
             _ ->  ... -- parse two records
                     return DebugEvent{ ...}
```